src/Nethermind/Nethermind.Core/Collections/SpanConcurrentDictionary.cs

```diff
@@ -0,0 +1,2191 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Nethermind.Core.Collections
+{
+    /// Adapted from .net source code.
+    /// <summary>Represents a thread-safe collection of keys and values.</summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    /// <remarks>
+    /// All public and protected members of <see cref="SpanConcurrentDictionary{TKey,TValue}"/> are thread-safe and may be used
+    /// concurrently from multiple threads.
+    /// </remarks>
+    [DebuggerTypeProxy(typeof(IDictionaryDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class SpanConcurrentDictionary<TKey, TValue> : IDictionary<TKey[], TValue>, IDictionary, IReadOnlyDictionary<TKey[], TValue> where TKey : notnull
```

Comment:
No tests for such a big class? 👁️

Reply:
While not perfect it is 50% covered: 

![obraz](https://user-images.githubusercontent.com/12445221/223546389-41c77ca7-fdb6-4bf2-9fca-a685f0ae3a86.png)

Was looking to bring tests from standard library but they are quite huge and incompatible.

Reply:
Thank you for providing the coverage @LukaszRozmej Have you considered the alternative with a struct like I described so that no additional collection is bring here? If yes, what what have you found? If not, why?

Reply:
I think we can consider those two things independent. Some of our interfaces are generic and based on `byte[]`, we could try packing them here and there to more precise ones, but that might not always be the case.